core/trino-main/src/main/java/io/trino/sql/planner/IrExpressionInterpreter.java

```diff
@@ -763,6 +765,24 @@ public Expression visitCast(Cast node, SymbolResolver context)
             };
         }
 
+        @Override
+        protected Expression visitArray(Array node, SymbolResolver context)
+        {
+            List<Expression> elements = node.elements().stream()
+                    .map(field -> processWithExceptionHandling(field, context))
+                    .toList();
+
+            if (elements.stream().allMatch(Constant.class::isInstance)) {
+                BlockBuilder builder = node.elementType().createBlockBuilder(null, node.elements().size());
+                for (Expression element : elements) {
+                    writeNativeValue(node.elementType(), builder, ((Constant) element).value());
+                }
+                return new Constant(node.type(), builder.build());
+            }
+
+            return new Array(node.elementType(), elements);
```

Comment:
If there are both constant and non-constant elements, does it make sense to represent it as the sum of arrays where one is `Constant` and second one is `Array` ? Does the order matter here?

Reply:
Yes, the order matters. In theory, we could find runs of constant and non-constant segments and split it, but I don’t think it buys us much.

Reply:
Sure, merge it then :)